src/CoreWCF.Primitives/src/CoreWCF/Channels/BinaryMessageEncoderFactory.cs

```diff
@@ -538,20 +536,20 @@ private ArraySegment<byte> ExtractSessionInformationFromMessage(ArraySegment<byt
                         }
 
                         int size = dictionarySize;
-                        int offset = messageData.Offset + headerSize;
+                        int offset = headerSize;
 
                         while (size > 0)
                         {
                             decoder.Reset();
-                            int bytesDecoded = decoder.Decode(buffer, offset, size);
+                            int bytesDecoded = decoder.Decode(messageData.Slice(offset, size));
                             int utf8ValueSize = decoder.Value;
                             offset += bytesDecoded;
                             size -= bytesDecoded;
                             if (utf8ValueSize > size)
                             {
                                 throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidDataException(SR.BinaryEncoderSessionMalformed));
                             }
-                            string value = Encoding.UTF8.GetString(buffer, offset, utf8ValueSize);
+                            string value = messageData.Slice(offset, utf8ValueSize).ParseAsUTF8String(utf8ValueSize);
                             offset += utf8ValueSize;
```

Comment:
We are told the size of the string in bytes (that's what the IntDecoder is returning) which is why you call `messageData.Slice(offset, utf8ValueSize)`, you end up with a ReadOnlySequence which represents the string and only the string. As this work is targeting CoreWCF 2.0, we can rely on newer api's. There's an extension method `EncodingExtensions.GetString` which takes a `ReadOnlySequence<byte>` and returns the string that was introduced in .NET 5. You can find the docs [here](https://docs.microsoft.com/en-us/dotnet/api/system.text.encodingextensions.getstring?view=net-5.0). That way you don't need your own ParseAsUTF8String implementation.

Reply:
Added a TODO so we can ctach thos once moved toCoreWCF2.0 https://github.com/CoreWCF/CoreWCF/pull/801/commits/d76d9b56a51f1ab211175beeec2004feac504ad3